### PR TITLE
chore: improve DX

### PR DESCRIPTION
**Pull request by ntn-x2**
Fixes https://github.com/KILTprotocol/ticket/issues/2974 and based on top of https://github.com/KILTprotocol/kilt-node/pull/574.

This is my take at trying to improve the DX of integrating DIP components. The following is a list of the main changes introduced by this PR:

- Definitions of traits that are only ever used and useful inside a pallet have been changed to take a generic `Runtime` type that must implement the pallet config, reducing the clutter when implementing them
- Additional bounds have been added to the associated types of those traits, so that they don't need to be specified in the pallet config so that we can have trait bounds of those types. So basically what used to be the trait bounds on the pallet config, is not a trait bound on the trait return type. I think this is also the "Rusty" way to go.
- The generic consumer types exported in `kilt-dip-support` crate have been renamed with the `Generic` prefix, because I introduced two new types, starting with `Kilt`, that make it easier to integrate any KILT runtime as an identity provider, with the downside of having to depend on the whole runtime definition, but there's also ways around that (e.g., by having an intermediate crate that only implements the required `Config` traits). Example 1 below shows the change.
- Better re-exports (feedback is welcome on this one). For instance, for types that implement a basic version of a pallet's trait definition, the type is exported directly from the pallet crate root, while the trait is still namespaced behind the `mod traits`. Example 2 below.
- Change of the origin check for the `pallet_dip_consumer::dispatch_as` extrinsic from `ensure_signed` to a generic `T::DispatchOriginCheck::ensure_origin`, as long as the `DispatchOriginCheck` returns an `AccountId` upon successful verification.

### Old

```rust
pub type ProofVerifier = VersionedDipSiblingProviderStateProofVerifier<
	RelayStateRootsViaRelayStorePallet<Runtime>,
	ConstU32<2_000>,
	ProviderParachainStateInfoViaProviderPallet<ProviderRuntime>,
	AccountId,
	BlakeTwo256,
	KeyIdOf<ProviderRuntime>,
	ProviderAccountId,
	Web3Name,
	LinkableAccountId,
	10,
	10,
	u128,
	// Signatures are valid for 50 blocks
	FrameSystemDidSignatureContext<Runtime, 50>,
	DipCallFilter,
>;
```

### New (100% equivalent to the old one)

```rust
pub type ProofVerifier = KiltVersionedSiblingProviderVerifier<
	ProviderRuntime,    // KILT runtime definition
	ConstU32<2_000>,
	RelayStateRootsViaRelayStorePallet<Runtime>,    // Local runtime definition
	BlakeTwo256,
	DipCallFilter,
	10,
	10,
	50,
>;
```

## Example 2: example of type implementing a pallet's trait with a basic implementation

For the trait

```rust
	pub trait IdentityCommitmentGenerator<Runtime>
	where
		Runtime: Config,
		Runtime::IdentityProvider: IdentityProvider<Runtime>,
	{
		type Error: Into<u16>;
		type Output: Clone + Eq + Debug + TypeInfo + FullCodec + MaxEncodedLen;

		fn generate_commitment(
			identifier: &Runtime::Identifier,
			identity: &IdentityOf<Runtime>,
			version: IdentityCommitmentVersion,
		) -> Result<Self::Output, Self::Error>;
	}
```

a basic implementation (mostly useful for tests) is provided

```rust
	pub struct DefaultIdentityCommitmentGenerator<Output>(PhantomData<Output>);

	impl<Runtime, Output> IdentityCommitmentGenerator<Runtime> for DefaultIdentityCommitmentGenerator<Output>
	where
		Runtime: Config,
		Output: Default + Clone + Eq + Debug + TypeInfo + FullCodec + MaxEncodedLen,
	{
		type Error = u16;
		type Output = Output;

		fn generate_commitment(
			_identifier: &Runtime::Identifier,
			_identity: &IdentityOf<Runtime>,
			_version: IdentityCommitmentVersion,
		) -> Result<Self::Output, Self::Error> {
			Ok(Output::default())
		}
	}
```

The pallet's crate will then export the trait with `pub mod traits`, while the type is re-exported directly _also_ from the root with `pub use traits::DefaultIdentityCommitmentGenerator`.

### CODE REVIEW

File: pallets/pallet-dip-consumer/src/traits.rs [hunk 1]
```diff
@@ -16,40 +16,42 @@
...
-	type Submitter;
+use crate::{Config, RuntimeCallOf};
+
+pub trait IdentityProofVerifier<Runtime>
+where
+	Runtime: Config,
+{
+	type Error: Into<u16>;
+	type Proof: Parameter;
 	type VerificationResult;
```
Comment (Ad96el) [inline pallets/pallet-dip-consumer/src/traits.rs:29]: This little change makes a lot of things much easier to follow

**Review by Ad96el — COMMENTED** [review]
LGTM. I'm considering adding a fourth runtime, where the required configuration traits for pallets are implemented. This would make integration easier for chains that want to use KILT as the provider. I can provide a more detailed review in a week when I have a deeper understanding of the DIP implementation. If necessary, a second review can follow. :)

File: pallets/pallet-dip-provider/src/traits.rs [hunk 2]
```diff
@@ -19,86 +19,103 @@
...
-pub mod identity_provision {
-	use sp_std::marker::PhantomData;
-
-	pub trait IdentityProvider<Identifier> {
-		type Error;
+		type Error: Into<u16>;
 		type Success;
 
-		fn retrieve(identifier: &Identifier) -> Result<Option<Self::Success>, Self::Error>;
+		fn retrieve(identifier: &Runtime::Identifier) -> Result<Option<Self::Success>, Self::Error>;
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/traits.rs:37]: The point is that we also combine the result of different identity providers together, and for instance the lack of a linked web3name is per se not an error. Hence it is up to the consumer of this information to decide whether a `None` is to be treated as an error or not. For instance, there could also be a default value if the result is a `None`, before being passed to the identity commitment generator.

File: crates/kilt-dip-support/src/utils.rs [hunk 3]
```diff
@@ -105,16 +105,17 @@ where
 	}
 }
 
-impl<Identifier, A, B, C> IdentityProvider<Identifier> for CombineIdentityFrom<A, B, C>
+impl<Runtime, A, B, C> IdentityProvider<Runtime> for CombineIdentityFrom<A, B, C>
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/utils.rs:108]: Yes, basically. I tried to still keep it generic but I see it does not make much sense. In that way, we could also probably remove the `Result<Option<T>>` from the trait and simply return an `Option<T>`. I will investigate that.

File: crates/kilt-dip-support/src/export/child.rs [hunk 4]
```diff
@@ -112,19 +118,161 @@ where
 	}
 }
 
-pub struct VersionedDipChildProviderStateProofVerifier<
+// Implements the same `IdentityProvider` trait, but it is internally configured
+// by receiving the runtime definitions of both the provider and the receiver.
+pub struct KiltVersionedChildProviderVerifier<
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/export/child.rs:123]: And?

File: pallets/pallet-dip-provider/src/traits.rs [hunk 5]
```diff
@@ -19,86 +19,103 @@
...
+		fn generate_commitment(
+			identifier: &Runtime::Identifier,
+			identity: &IdentityOf<Runtime>,
+			version: IdentityCommitmentVersion,
+		) -> Result<Self::Output, Self::Error>;
+	}
+
+	// Implement the `IdentityCommitmentGenerator` by returning the `Default` value
+	// for the `Output` type.
+	pub struct DefaultIdentityCommitmentGenerator<Output>(PhantomData<Output>);
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/traits.rs:99]: We had a conversation once with Albi about this. We agreed that `()` does not convey enough information about what it does. I could add it here, but I would still like to have proper structs being exported here, as they are clear in what they do. Nowadays, if you use rust-analyzer in VSCode, hovering over a trait tells you what types implement it, which will make discovery easier.

**Comment by ntn-x2** [comment]
> LGTM. I'm considering adding a fourth runtime, where the required configuration traits for pallets are implemented. This would make integration easier for chains that want to use KILT as the provider.

If you mean so that a consumer can call the Kilt-flavored struct to implement the `ProofVerifier` trait, then yes, that makes sense, I mentioned in the third point of the bullet list above. That would probably be exported inside the `kilt-dip-support` crate, in its own module and re-exported from the root. If you mean something else, then I don't think I understood your point.

File: crates/kilt-dip-support/src/export/child.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/export/child.rs:123]: Oh you mean between the two. One is to be deployed on a sibling parachain, one is to be deployed on the relaychain. The second expects one more element in the proof, which is the header of the block for which the state proof is being presented.

File: pallets/pallet-dip-provider/src/traits.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/traits.rs]: I fixed it in https://github.com/KILTprotocol/kilt-node/pull/577/commits/1e338a414a8bf1a170f552036d42240fc101e440. No need to have an `Option` inside a `Result`.

File: crates/kilt-dip-support/src/utils.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/utils.rs]: Fixed in https://github.com/KILTprotocol/kilt-node/pull/577/commits/1e338a414a8bf1a170f552036d42240fc101e440.

File: crates/kilt-dip-support/src/export/child.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply crates/kilt-dip-support/src/export/child.rs:123]: This is logic that does not involve storing anything on chain, hence more tweaks can be applied later on, if people find these hard to deal with.

File: pallets/pallet-dip-consumer/src/lib.rs [hunk 6]
```diff
@@ -26,6 +26,7 @@ pub mod traits;
 mod origin;
 
 pub use crate::{origin::*, pallet::*};
+pub use traits::SuccessfulProofVerifier;
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-consumer/src/lib.rs]: https://github.com/KILTprotocol/kilt-node/pull/577/commits/467d8df73ac974ec6dab2516f306a7e5b4449ecc

File: dip-template/runtimes/dip-provider/src/dip.rs [hunk 7]
```diff
@@ -48,8 +48,7 @@ pub mod runtime_api {
 
 	#[derive(Encode, Decode, TypeInfo)]
 	pub enum DipProofError {
-		IdentityNotFound,
-		IdentityProviderError(<LinkedDidInfoProviderOf<Runtime> as IdentityProvider<DidIdentifier>>::Error),
+		IdentityProviderError(LinkedDidInfoProviderError),
```
Comment (Ad96el) [inline dip-template/runtimes/dip-provider/src/dip.rs:51]: These errors are not following our convention not to have the term 'Error' in the suffix

File: pallets/pallet-dip-provider/src/traits.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-dip-provider/src/traits.rs:84]: Ah, okay, that makes sense. If you've already discussed it with someone, feel free to ignore my comment. It's probably just a matter of taste

File: crates/kilt-dip-support/src/export/child.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply crates/kilt-dip-support/src/export/child.rs:123]: Ah, okay, so the dip can also be used in the relay chain. Makes sense.

**Review by Ad96el — APPROVED** [review]
I would likely review it once more, if I have the time, but I believe it's ready to be merged

**Review by trusch — APPROVED** [review]
This looks nice, especially the Kilt specific types will make it super easy for consumers to integrate with us without loosing the generic types that would work with any provider!

**Pull request merged**